### ISSUE TITLE
Decompose IMS Content Packages (SCORM/IMSCP) into individual Kolibri resources

### ISSUE THREAD

**Issue by rtibbles** [learningequality/ricecooker#332]
## Overview

Migrate the `imscp` code into ricecooker.

Decompose an IMS Content Package (including SCORM) into individual Kolibri resources, driven by its manifest.

This reverses the original "no dependency zips" decision; shared-library dependency zips return as a follow-up.

**Complexity:** High
**Target branch:** main

## Context

- The `imscp` code lives in a separate repo (`learningequality/imscp`) and should be native to ricecooker.
- IMSCP is the packaging layer under SCORM.
- Its manifest (`imsmanifest.xml`) is the standardized, tool-independent structure.
- One opaque webcontent zip backs many nodes, so Kolibri's per-resource import pulls the whole package when a single node is selected.
- Wrapping plain media in an HTML5 sandbox strips native affordances — a video should play as a VideoNode, not a `<video>` in a sandbox.
- The existing reference machinery locates and rewrites references only within a single archive.

## The decomposition ladder

Per manifest resource, take the highest rung it safely qualifies for:

1. Native media node (video/audio/pdf/image) — a single media file in a wrapper.
2. KPUB — static article/document HTML.
3. HTML5 zip served via zipcontent — genuinely interactive content.
4. QTI assessment item — assessment content.

Classification runs on one axis: does the resource's meaning survive stripping its JS/CSS?

- Discount SCORM API boilerplate (`pipwerks`, `LMSInitialize`, `SetValue`) before judging interactivity.
- Bias conservative — promote only on high confidence.
- Never promote a resource carrying assessment, score, or tracking semantics.

## The crux

Decomposition must rewrite references *across* resources: a promoted node referenced by a sibling, or a shared library in a dependency zip.

The rewrite emits two new target forms:

- native-node references
- zipcontent-relative dependency paths

## Deferred to follow-ups

- Per-tool deep parsers (Storyline, Captivate, Rise, iSpring, Adapt content models).
- Interactivity classification beyond the sanitization-survival test.
- Automatic dependency-zip vendoring for shared HTML5 libraries.

## Acceptance Criteria

- [ ] An IMSCP/SCORM package imports natively, with the `imscp` code migrated into ricecooker.
- [ ] The manifest drives the node/topic tree.
- [ ] A single wrapped media file becomes a native media node.
- [ ] A static-article resource becomes a KPUB.
- [ ] An interactive resource becomes an HTML5 zip served via zipcontent.
- [ ] A resource carrying assessment semantics is never promoted to media or KPUB.
- [ ] References between decomposed resources resolve correctly in Kolibri.
- [ ] Each decomposed node is backed by its own files, not one shared package zip.

**Comment by rtibblesbot** [comment]
Watching for resolution on behalf of learningequality/ricecooker#337.

**Comment by rtibblesbot** [comment]
## Design review — 4 blocking questions before implementation

I ran the brainstorming pass against the current codebase (pipeline, `archive_assets`, `references.py`) and the legacy `learningequality/imscp` source. The legacy code parses `imsmanifest.xml` into a topic tree but **always wraps every resource as an `HTML5AppNode`** (whole-package zip, or an entrypoint redirect + `HTML5_DEPENDENCY_ZIP`). This issue asks for something substantially new — the decomposition ladder — so four forks below change what gets built. Each has a recommended default; confirming them (or correcting) unblocks a plan.

**1. Where does decomposition live?**
The pipeline is single-file → single-node (one `File` → one node's files). IMSCP decomposition explodes one package into a *heterogeneous node subtree* (topics + media/KPUB/HTML5/QTI leaves). That does not fit the "new file type = 2 pipeline changes" model, and the pipeline has no concept of emitting a subtree.
*Recommend:* a chef-facing utility (successor to legacy `make_topic_tree`, e.g. `ricecooker/utils/imscp.py`) that a `SushiChef` calls to expand an IMSCP zip into a node tree, reusing the pipeline per emitted leaf file. **Not** a `convert.py` handler. Confirm?

**2. Scope boundary with #337 (QTI).**
Rung 4 is "QTI assessment item," but #337 ("Handle QTI exercises") owns per-item `type='QTI'` ingestion (raw_data XML via the internal API, per studio#6002). le_utils only exposes a `QTI_ZIP` preset today, not per-item.
*Recommend:* #332 **detects** assessment/score/tracking semantics and refuses to promote them (keeps them as HTML5 zips), satisfying "never promoted to media or KPUB." Actual `type='QTI'` item emission is deferred to #337. Confirm #332 does not emit QTI items itself?

**3. KPUB from static HTML.**
No existing path converts an IMSCP HTML resource (HTML + its referenced assets) into a valid KPUB. `DocumentConversionHandler` runs pandoc on *documents* (docx/odt/rtf/md), not HTML-with-assets; `KPUBConversionHandler` only validates/sanitizes an already-`.kpub` archive.
*Recommend:* which is in scope — (a) build a new HTML-article → KPUB packager, (b) pandoc HTML → KPUB, or (c) ship static articles as sanitized HTML5-article zips for now and defer true KPUB generation? This is a large swing in effort.

**4. Cross-node reference target form.**
AC "references between decomposed resources resolve correctly in Kolibri" needs a concrete rewrite target. There is **no** node-to-node reference convention in ricecooker today (no `content://`-style scheme), and `archive_assets`/`references.py` only rewrite external URLs and within-archive paths.
*Recommend:* what exact string should a reference to a promoted sibling rewrite to, and does Kolibri/zipcontent resolve it today? If the scheme doesn't exist yet, this AC can't be fully delivered here and cross-node promotion may need to stay conservative (keep mutually-referencing resources together as one HTML5 zip) until the scheme lands.

**Assumptions I'll proceed on if unstated** (flag if wrong):
- Classifier operationalizes "meaning survives stripping JS/CSS" as: discount SCORM boilerplate (`pipwerks`/`LMSInitialize`/`SetValue`), run `strip_scripts`, then promote only when the residual is a single media element (→ media) or static prose/media with no interactive residue (→ KPUB/HTML article); everything else stays an HTML5 zip. Bias conservative.
- Test fixtures migrated from the legacy repo's `examples/` (`gitta_ims.zip`, `eventos.zip`, `test_quiz.zip`); pipeline tests land in `tests/pipeline/`.

Answering **1–4** is enough to write the implementation plan.

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibbles** [comment]
1. Where does decomposition live?

The pipeline handler _already_ has support for emitting multiple files from a stage - we are now expanding this to add support for emitting multiple ContentNodes. In the original implementation there was a special kind of ContentNode that could flexibly become a Topic/Folder as needed - we should be allowing ContentNodes to do this too, if the file they are handed returns a tree representation for metadata. The decomposition still fully lives in the pipeline, with the existing file based branching logic extended to the content node metadata also.

2. Scope boundary with https://github.com/learningequality/ricecooker/issues/337 (QTI).

I think we can go a step further and just reject them for now. This can then be changed in the follow up issue.

3. KPUB from static HTML

This is much simpler than you are making out - pandoc already converts docx to an HTML zip - what we are doing is merely adding a layer that takes any arbitrary HTML5 zip and turns it into a KPUB if it meets the criteria. So, this would just become part of the HTML5 zip conversion handler - in some cases, it will emit a KPUB instead!

4. Cross-node reference target form

Let's just not implement this at all. I will file follow up issues for this.

**Comment by rtibblesbot** [comment]
## Resolved design — sufficiently specified to plan

Thanks @rtibbles — all four forks answered. Confirmed each against the code; the direction fits the existing pipeline cleanly. Consolidated design for the implementation plan:

**1. Decomposition lives in the pipeline (not a chef-side utility).**
The pipeline already carries content-node intent alongside files: a handler returns `FileMetadata(content_node_metadata=ContentNodeMetadata(kind=…, extra_fields=…))`, and `extract_metadata` already sets `kind` from the file (`context.py`, `extract_metadata.py:42`). The extension: let `ContentNodeMetadata` carry a **tree** (children), and let a `ContentNode` expand into a Topic/Folder subtree when handed a file that yields one — the same "flexible node becomes Topic or leaf" behavior the legacy code had, now driven by content-node metadata rather than a bespoke node class. An IMSCP handler parses `imsmanifest.xml` (migrating `learningequality/imscp` `core.py`'s manifest walk) and emits that tree; each leaf's own file re-enters the pipeline for its kind/preset.

**2. Assessment/QTI resources are rejected, not ingested.**
Detect assessment/score/tracking semantics (discount SCORM boilerplate — `pipwerks`/`LMSInitialize`/`SetValue` — first) and drop those resources for now. Per-item `type='QTI'` emission is #337's job.

**3. KPUB is a branch inside the HTML5 zip conversion handler.**
No new packager. `HTML5ConversionHandler` already returns content-node metadata; extend it so a sanitized HTML5 zip that meets KPUB criteria is emitted as a KPUB (`HTML5_ARTICLE`/`KPUB_ZIP`) instead of `HTML5_ZIP`. Pandoc already yields HTML zips upstream, so this "promote HTML5 → KPUB when it qualifies" layer covers the static-article rung.

**4. Cross-node reference rewriting is out of scope.**
Not implemented here; @rtibbles will file follow-ups. Correspondingly descoped: the "crux" section and the AC *"References between decomposed resources resolve correctly."* Where resources reference each other, stay conservative (keep them together as one HTML5 zip) rather than promote and break a link.

**Decomposition ladder as built by #332:** native media node → KPUB (via the HTML5 handler branch) → HTML5 zip → **reject** (assessment). Classifier stays conservative: promote only on the sanitization-survival test (single media element → media; static prose/media with no interactive residue → KPUB), else HTML5 zip.

**Also in scope:** migrate the `imscp` manifest parser and its SCORM fixtures (`gitta_ims.zip`, `eventos.zip`, `test_quiz.zip`) into ricecooker; pipeline tests land in `tests/pipeline/`.

**AC status after descoping:** "imports natively / migrated" ✓, "manifest drives tree" ✓, "wrapped media → native node" ✓, "static article → KPUB" ✓, "interactive → HTML5 zip" ✓, "assessment never promoted" ✓ (via rejection), "each node backed by its own files" ✓. *"References resolve correctly"* → deferred to a follow-up.

Proceeding to write the implementation plan.

**Comment by rtibblesbot** [comment]
**Plan:** Migrate the `learningequality/imscp` manifest parser into ricecooker and decompose an IMS Content Package (incl. SCORM) into a native Kolibri node subtree — topics from the manifest, and per-resource media / KPUB / HTML5 leaves — driven entirely by the file-processing pipeline.

- [ ] Migrate the IMSCP manifest parser
- [ ] Tree fields on `ContentNodeMetadata`
- [ ] Expand a metadata tree into descendant nodes
- [ ] SCORM-boilerplate discount + assessment / single-media classifiers
- [ ] KPUB-promotion branch in `HTML5ConversionHandler`
- [ ] `IMSCPConversionHandler` — decompose the package into the node tree

---
_@rtibblesbot's comments are generated by an LLM, and should be evaluated accordingly_

<details><summary>How was this generated?</summary>

- Implemented code following the pre-approved plan using test-driven development (red-green-refactor)
- Ran the project's full test suite and linter before committing
- For UI changes: captured live screenshots to verify rendering
- For backend changes: verified with example API/CLI output

</details>

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibblesbot** [comment]
The bot has been waiting for a response for 2 hour(s). It will resume automatically when someone replies.

**Comment by rtibbles** [comment]
Fixed in #721
